src/primitives.rs

```diff
@@ -0,0 +1,161 @@
+
+pub trait Message {
+    fn get_size(&self) -> i32;
+}
+
+pub trait FromByte: Default {
+    fn decode(&mut self, buf: &mut Buf) -> Result<(), DecodeErrors>;
+
+    fn decode_new(buf: &mut Buf) -> Result<Self, DecodeErrors> {
+        let mut temp = Self::default();
+        match temp.decode(buf) {
+            Ok(_) => Ok(temp),
+            Err(e) => Err(e),
+        }
+    }
+}
+
+pub trait DecodableMessage: Message + FromByte {}
+
+pub trait Request: DecodableMessage {
+    fn get_header(&self) -> &HeaderRequest;
+}
+
+// Skip for now
+pub trait Response: DecodableMessage {}
+
+impl FromByte for i8 {
+    fn decode(&mut self, buf: &mut Buf) -> Result<(), DecodeErrors> {
+        *self = buf.get_i8();
+        Ok(())
+    }
+}
+
+impl FromByte for i16 {
+    fn decode(&mut self, buf: &mut Buf) -> Result<(), DecodeErrors> {
+        *self = buf.get_i16_be();
+        Ok(())
+    }
+}
+
+impl FromByte for i32 {
+    fn decode(&mut self, buf: &mut Buf) -> Result<(), DecodeErrors> {
+        *self = buf.get_i32_be();
+        Ok(())
+    }
+}
+
+impl FromByte for String {
+    fn decode(&mut self, buf: &mut Buf) -> Result<(), DecodeErrors> {
+        let mut length: i16 = 0;
+        length.decode(buf);
+        if length == 0 || length == -1 {
+            return Ok(());
+        }
+        // Length should not be smaller than -1 according to the Kafka protocol
+        if length < -1 {
+            // ToDo: throw error
+            return Ok(());
+        }
+        self.reserve(length as usize);
+        // Any smart way to read a String from buffer?
+        let mut temp_vec: Vec<u8> = vec![0; length as usize];
+        buf.copy_to_slice(temp_vec.as_mut_slice());
+        // ToDo: throw error
+        *self = String::from_utf8_lossy(temp_vec.as_ref()).parse().unwrap();
+
+        if self.len() != length as usize {
+            // ToDo: throw error
+            return Ok(());
+        }
+        Ok(())
+    }
+}
+
+// Reference: https://github.com/spicavigo/kafka-rust/blob/c58cf5f30b35fad6ab163416d51d2b99a30da9c2/src/protocol/mod.rs#L108
+#[derive(Debug)]
+pub struct HeaderRequest {
+    pub api_key: i16,
+    pub api_version: i16,
+    pub correlation_id: i32,
+    pub client_id: String,
+}
+
+impl HeaderRequest {
+    fn new(
+        api_key: i16,
+        api_version: i16,
+        correlation_id: i32,
+        client_id: String,
+    ) -> HeaderRequest {
+        HeaderRequest {
+            api_key,
+            api_version,
+            correlation_id,
+            client_id,
+        }
+    }
+}
+
+impl std::fmt::Display for HeaderRequest {
+    fn fmt(&self, f: &mut Formatter) -> Result<(), Error> {
+        println!(
+            "api_key: {}, api_version: {}, correlation_id: {}, client_id: {}",
+            self.api_key, self.api_version, self.correlation_id, self.client_id
+        );
+        Ok(())
+    }
+}
+
+impl Default for HeaderRequest {
+    fn default() -> HeaderRequest {
+        Self::new(-1, -1, -1, String::new())
```

Comment:
Is this specified in the Kafka spec, or is `-1` chosen as an invalid value? If the case is that this is invalid and is intended to be set later, I'd _strongly_ advise against this. Idiomatic `Default` impls should typically construct valid instances of a type, and this seems somewhat error prone to me.

I think getting rid of `decode_new` and having `decode` return `Result<Self, ...>` will obviate the need for this.

Reply:
I'll try to remove `decode_new` and make `decode` return `Result<Self, ...>`.